### PR TITLE
feat(audiobook): sync progress to Yandex, surface in search and browse

### PR DESCRIPTION
**Pull request by trudenboy**
## Summary

Closes three visible gaps left after chapter-concatenation streaming (v3.2.0) — the P0 items from the recent audiobook audit.

- **Progress sync.** `on_played` / `on_streamed` now push the current chapter position via a new `api_client.play_audio` wrapper, so Yandex's other clients see the same resume point. `parse_audiobook` propagates `album.listening_finished` → `Audiobook.fully_played`.
- **Search.** `MediaType.AUDIOBOOK` maps to Yandex `"album"` (no dedicated type on their side) and the albums bucket is split via `classify_album` into albums/audiobooks. `requested_types` is deduped so `ALBUM + AUDIOBOOK` still issues a single API call.
- **Browse.** Collection now renders *My Audiobooks* / *My Podcasts* folders (RU + EN). Subpaths are added to `_known_folders` so the base `MusicProvider.browse` routes them to the existing `get_library_audiobooks` / `get_library_podcasts` generators.

### Noteworthy implementation details

- **Chapter-id cache.** `MediaItemChapter` is `frozen=True` without a `chapter_id` field, so per-audiobook `(chapter_ids, chapter_durations_ms)` is kept in `self._audiobook_chapter_cache` (populated in `_get_audiobook_stream_details`, with a fallback to `get_album_with_tracks` for cold `on_played` hits).
- **Per-session play_id.** `uuid4` minted on first `on_played` per audiobook, reused across the 30-sec ticks, cleared in `on_streamed`.
- **Reverse mapping.** The existing `_resolve_audiobook_seek` is reused to turn global position (seconds) back into `(chapter_idx, offset)` for `play_audio`.
- **Errors are advisory.** `api_client.play_audio` swallows `BadRequestError / NetworkError / ProviderUnavailableError` — a failed progress push must never abort pause/stop.
- **`get_resume_position` left unimplemented.** Per-chapter server position isn't exposed by the yandex-music SDK; MA falls back to its internal tracker. `fully_played` is still surfaced on the Audiobook model.

## Test plan

- [x] 13 new unit tests covering: `fully_played` extraction, `on_played/on_streamed` progress reporting, play_id reuse across ticks, radio-branch isolation, search bucket split + dedup, Collection folder rendering for RU/EN locales.
- [x] Full suite green: `206 passed` via the sibling ynison venv (see `memory/reference_test_venv.md` for the exact PYTHONPATH invocation).
- [x] `ruff check` clean.
- [x] `mypy` pre-commit hook passes (with `MYPYPATH=... ma-server`).
- [ ] Manual e2e: play an audiobook in MA, check the Yandex mobile app mirrors the position; confirm "My Audiobooks" folder and audiobook search cards appear.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: provider/provider.py [hunk 1]
```diff
@@ -3059,3 +3141,57 @@ async def on_streamed(self, streamdetails: StreamDetails) -> None:
...
+        play_id = self._audiobook_play_ids.setdefault(audiobook_id, uuid.uuid4().hex)
+        track_length_sec = chapter_durations_ms[idx] // 1000
+        await self.client.play_audio(
+            track_id=chapter_ids[idx],
+            album_id=audiobook_id,
+            play_id=play_id,
+            track_length_seconds=track_length_sec,
+            total_played_seconds=int(offset),
+            end_position_seconds=int(offset),
+        )
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py:3232]: Done in 7c64a51 — `_report_audiobook_progress` now wraps `_resolve_audiobook_chapter_map` in `try/except ResourceTemporarilyUnavailable`, logs at debug, and returns without pushing progress. Matches the existing best-effort contract.

File: provider/provider.py [hunk 2]
```diff
@@ -150,6 +151,11 @@ class YandexMusicProvider(MusicProvider):
     # that all derive from the same liked-albums endpoint.
     _liked_albums_cache: tuple[float, list[YandexAlbum]] | None = None
     _liked_albums_lock: asyncio.Lock
+    # Per-audiobook cache of (chapter_track_ids, chapter_durations_ms) used to
+    # report playback progress per chapter via play_audio.
+    _audiobook_chapter_cache: dict[str, tuple[list[str], list[int]]]
+    # Stable play_id per audiobook session, cleared in on_streamed.
+    _audiobook_play_ids: dict[str, str]
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py:173]: Done in 7c64a51 — `_report_audiobook_final` pops the cache entry for the finished audiobook, and `unload()` clears both dicts. Working set stays at O(currently-playing audiobooks), no tunable LRU needed.

File: provider/provider.py [hunk 3]
```diff
@@ -2758,6 +2803,30 @@ def _resolve_audiobook_seek(
...
+        album = await self.client.get_album_with_tracks(audiobook_id)
+        if not album or not (album.volumes or []):
+            return [], []
+        chapter_ids: list[str] = []
+        chapter_durations_ms: list[int] = []
+        for disc in album.volumes or []:
+            for track_obj in disc:
+                chapter_ids.append(str(track_obj.id))
+                chapter_durations_ms.append(int(track_obj.duration_ms or 0))
+        self._audiobook_chapter_cache[audiobook_id] = (chapter_ids, chapter_durations_ms)
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py:2834]: Done in 7c64a51 — extracted `_extract_chapter_map_from_album(album)` at module level; `_get_audiobook_stream_details` and `_resolve_audiobook_chapter_map` both use it now.

File: provider/api_client.py [hunk 4]
```diff
@@ -337,6 +337,40 @@ async def _send(c: ClientAsync) -> bool:
...
+                        album_id=album_id,
+                        from_=from_,
+                        play_id=play_id,
+                        track_length_seconds=track_length_seconds,
+                        total_played_seconds=total_played_seconds,
+                        end_position_seconds=end_position_seconds,
+                    )
+                )
+            )
+        except (BadRequestError, NetworkError, ProviderUnavailableError) as err:
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/api_client.py]: Done in 7c64a51 — `play_audio` now also catches `UnauthorizedError`, `LoginFailed` and `ResourceTemporarilyUnavailable`. Kept the list explicit instead of a blanket `Exception` to stay consistent with the rest of `api_client.py`.

File: provider/provider.py [hunk 5]
```diff
@@ -3032,11 +3114,18 @@ def _ensure_dont_stop_the_music_for_queue(self, queue_id: str | None) -> None:
...
-        can improve recommendations.
+        For radio: sends trackFinished or skip with actual seconds_streamed so
+        Yandex can improve recommendations.
+
+        For audiobooks: sends a final play_audio with the absolute stream position
+        so the last listening point is preserved in Yandex.
         """
+        data = streamdetails.data if isinstance(streamdetails.data, dict) else None
+        if streamdetails.media_type == MediaType.AUDIOBOOK and data and "chapter_ids" in data:
+            await self._report_audiobook_final(streamdetails, data)
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py]: Done in ce8a21c — `on_streamed` now routes on `media_type == MediaType.AUDIOBOOK` alone and calls `_report_audiobook_final(streamdetails, data or {})`. The cache pop + `play_id` pop happen before the early-return, so cleanup is guaranteed even when `data` is missing/stripped. Covered by two new tests (`test_on_streamed_audiobook_without_data_still_cleans_up`, `test_on_streamed_audiobook_branch_taken_even_without_chapter_data`).

File: provider/provider.py [hunk 6]
```diff
@@ -3059,3 +3148,76 @@ async def on_streamed(self, streamdetails: StreamDetails) -> None:
...
+        chapter_ids = data.get("chapter_ids") or []
+        chapter_durations_ms = data.get("chapter_durations_ms") or []
+        play_id = self._audiobook_play_ids.pop(audiobook_id, None) or uuid.uuid4().hex
+        self._audiobook_chapter_cache.pop(audiobook_id, None)
+        if not chapter_ids or not chapter_durations_ms:
+            return
+        absolute_sec = int(streamdetails.seek_position + (streamdetails.seconds_streamed or 0))
+        idx, offset = self._resolve_audiobook_seek(
+            chapter_durations_ms, max(absolute_sec, 0), len(chapter_ids)
+        )
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py]: Done in ce8a21c — extracted `_audiobook_progress_point` helper that maps `absolute_sec >= total_duration` to `(last_idx, last_chapter_duration_sec)` so natural EOF reports the end of the last chapter, not its start. Used from both `_report_audiobook_progress` and `_report_audiobook_final`. New test `test_on_streamed_audiobook_reports_end_of_last_chapter_at_eof` pins the behaviour.

File: provider/provider.py [hunk 7]
```diff
@@ -3059,3 +3148,76 @@ async def on_streamed(self, streamdetails: StreamDetails) -> None:
...
+            chapter_durations_ms, int(max(position_sec, 0)), len(chapter_ids)
+        )
+        play_id = self._audiobook_play_ids.setdefault(audiobook_id, uuid.uuid4().hex)
+        track_length_sec = chapter_durations_ms[idx] // 1000
+        await self.client.play_audio(
+            track_id=chapter_ids[idx],
+            album_id=audiobook_id,
+            play_id=play_id,
+            track_length_seconds=track_length_sec,
+            total_played_seconds=int(offset),
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py]: Done in ce8a21c — the same `_audiobook_progress_point` helper clamps `track_length_seconds` to at least 1 and `offset` to `[0, track_length_seconds]`, so a chapter with missing/zero `duration_ms` no longer ships `track_length_seconds=0` to Yandex. New test `test_on_streamed_audiobook_clamps_zero_duration_chapter` guards this.

File: provider/provider.py [hunk 8]
```diff
@@ -3059,3 +3152,109 @@ async def on_streamed(self, streamdetails: StreamDetails) -> None:
...
+        swallowed here in addition to the errors already absorbed inside
+        ``api_client.play_audio``.
+        """
+        try:
+            chapter_ids, chapter_durations_ms = await self._resolve_audiobook_chapter_map(
+                audiobook_id
+            )
+        except ResourceTemporarilyUnavailable as err:
+            self.logger.debug(
+                "Skipping audiobook progress report for %s (upstream unavailable): %s",
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py]: Done in 5f865ae — `_report_audiobook_progress` now re-raises `asyncio.CancelledError` and swallows every other `Exception` from chapter-map resolution. `api_client.play_audio` already has its own broad except, so the whole advisory path is now robust to auth / retry / network edge cases. +2 tests pin the behaviour (generic Exception swallowed, CancelledError still propagated).

File: provider/provider.py [hunk 9]
```diff
@@ -3059,3 +3152,109 @@ async def on_streamed(self, streamdetails: StreamDetails) -> None:
...
+        - At/beyond end-of-book, map to end of the last chapter (not start),
+          so Yandex's resume point doesn't rewind to the start of the final
+          chapter on natural completion.
+        - ``track_length_seconds`` is clamped to at least 1 and ``offset`` to
+          ``[0, track_length_seconds]`` — a chapter with ``duration_ms=None``
+          (coerced to 0 by the chapter-map builder) would otherwise send
+          ``track_length_seconds=0`` and block progress from syncing.
+        """
+        absolute_sec = max(0, absolute_sec)
+        total_duration_sec = sum(chapter_durations_ms) // 1000
```
[the remark this answers is not in the document]
Comment (trudenboy) [reply provider/provider.py:3177]: Skipping this one — `sum()` over a ~100-element int list is ~microseconds, `_audiobook_progress_point` runs at most once per 30s playback tick, and the suggested cache keyed on `id(chapter_durations_ms)` would be brittle (fresh list on each `get_album_with_tracks` call invalidates it) and reintroduces the exact unbounded-growth pattern we fixed in the previous round. Happy to revisit if profiling shows it actually matters.
